### PR TITLE
allow executable schema from data source

### PR DESCRIPTION
**Pull request by ecwyne**
Per discussion in https://github.com/gramps-graphql/gramps-express/issues/39

A data source can either provide
```
typeDefs: String
resolvers: Object
```
or
```
schema: GraphQLSchema
```

This implementation is extremely generous, flexible enough for both the former definition of a data-source and the new at the same time.

### CODE REVIEW

**Comment by jlengstorf** [comment]
Can you remind me: what was the use case for needing this?

I'm a little concerned that we're adding functionality that takes us away from the goal of creating a predictable API for data sources. If a data source can effectively be arbitrary code, what value does GrAMPS add over simply using schema stitching?

The conversation on https://github.com/gramps-graphql/gramps-express/issues/39 got a little broad, so forgive me if we already covered this and I'm not remembering.

**Comment by ecwyne** [comment]
The most complete arguments are [here](https://github.com/gramps-graphql/gramps-express/issues/39#issuecomment-340615131) and [here](https://github.com/gramps-graphql/gramps-express/issues/39#issuecomment-341142894)

I would argue that if the structure of a data-source is **too** constrained then the vision of having [all data exposed as graphql data sources](https://code.lengstorf.com/presentations/graphql-microservices/slides/#/56) will never be possible. If a data-source is created using the provided structure from data-source-base then an author can gain lots of additional features from gramps out of the box. But they shouldn't be precluded from exploring other patterns (especially if their use case doesn't map well to the prescribed pattern).

File: src/gramps.js [hunk 1]
```diff
@@ -33,23 +33,27 @@ const getDefaultApolloOptions = options => ({
...
   sources
-    .map(({ schema: typeDefs, resolvers, mocks, namespace }) => {
-      if (!typeDefs) {
+    .map(({ schema, typeDefs, resolvers, mocks, namespace }) => {
+      let executableSchema;
+      if (schema instanceof GraphQLSchema) {
+        executableSchema = schema;
+      } else if (typeof schema === 'string' || typeof typeDefs == 'string'){
+        executableSchema = makeExecutableSchema({
+          typeDefs: typeDefs || schema,
```
Comment (jlengstorf) [inline src/gramps.js]: I think my primary discomfort with this all balls up here. This makes it hard to write docs: "you can provide a `schema` defenition as described in [the GraphQL docs](http://graphql.org/learn/schema/) (which you can also call `typeDefs`), or you can provide a `schema` as generated by [`makeExecutableSchema`](https://www.apollographql.com/docs/graphql-tools/generate-schema.html#makeExecutableSchema), which requires that you supply the GraphQL schema and resolvers to `makeExecutableSchema` inside your data source — but don't worry, we'll do that for you."

That's hard to follow, and someone without a reasonably strong contextual knowledge of both GraphQL's _and_ Apollo's semantics will be completely lost reading these docs. We're effectively asking someone to read two other sets of docs before they can read our docs (or we're duplicating two sets of docs to keep them on-page, which is a bad idea).

The original intention was to limit the number of abstractions someone needed to learn before creating a simple data source. And while we can do a simple quickstart to get around that, we'll end up with the same problem the existing Apollo server docs have, which is that they have a lot of edge case functions that aren't documented with examples, and effectively just let you know they exist and then require that you look at the source and/or just try things to see what the errors look like.

Comment (ecwyne) [reply src/gramps.js]: My desired API would allow for **EITHER** (`typeDefs` + `resolvers`) **OR** (in *advanced cases*) `schema`

The main reason I even bring this feature up before a 1.0 is because it draws a distinction between `typeDefs` and `schema`

If a 1.0 release used `typeDefs` instead of `schema` I would be happy enough because it's **trivially easy** to allow for data sources to provide an executable `schema` for advanced cases (like remote schemas)

Comment (jlengstorf) [reply src/gramps.js]: Alright — let's do it. Swap out `schema` for `typeDefs`, and let's drop support for executable schemas. I still think it's confusing that we talk about GraphQL schemas, but then refer to them as `typeDefs` in code, but I'll survive.

If we bump up against a use case that requires exposing an executable schema, we can revisit this, but I'd prefer to avoid `OR` options in our config unless there's a _really_ good reason to do it.

**Comment by jlengstorf** [comment]
The comment thread was buried with 03174bf, so copying here for easier visibility:

> Alright — let's do it. Swap out schema for typeDefs, and let's drop support for executable schemas. I still think it's confusing that we talk about GraphQL schemas, but then refer to them as typeDefs in code, but I'll survive.
> 
> If we bump up against a use case that requires exposing an executable schema, we can revisit this, but I'd prefer to avoid OR options in our config unless there's a really good reason to do it.

**Comment by jlengstorf** [comment]
Closing this in favor of #19
